File: Contacts/Services/ContactsService.cs
using Contacts.DTOs;
using Contacts.Interfaces;
using Contacts.Model;
using System.Net.Mail;

namespace Contacts.Services
{
    public class ContactsService : IContactsService
    {
        private readonly IContactsRepository _contactsRepository;
        private readonly ICategoriesRepository _categoriesRepository;
        private readonly ISubCategoriesRepository _subcategoriesRepository;

        public ContactsService(
            IContactsRepository contactsRepository,
            ICategoriesRepository categoriesRepository,
            ISubCategoriesRepository subcategoriesRepository)
        {
            _contactsRepository = contactsRepository;
            _categoriesRepository = categoriesRepository;
            _subcategoriesRepository = subcategoriesRepository;
        }

        public Guid Create(CreateContactDTO contactDto)
        {
            try
            {
                var contact = ContactDTO.ToEntity(
                    new ContactDTO
                    {
                        Name = contactDto.Name,
                        Surname = contactDto.Surname,
                        Email = contactDto.Email,
                        Password = contactDto.Password,
                        CategoryId = contactDto.CategoryId,
                        SubcategoryId = contactDto.SubcategoryId,
                        CustomSubCategory = contactDto.CustomSubCategory,
                        Phone = contactDto.Phone,
                        BirthDate = contactDto.BirthDate
                    }
                    );
                if (contact == null)
                {
                    return Guid.Empty;
                }
                _contactsRepository.Create(contact);
                return contact.Id;
            }
            catch (ArgumentException) { return Guid.Empty; }
            catch (KeyNotFoundException) { return Guid.Empty; }
        }

        public bool Delete(Guid id)
        {
            try
            {
                _contactsRepository.Delete(id);
                return true;
            }
            catch (KeyNotFoundException)
            {
                return false;
            }
        }

        public ICollection<ContactDTO> GetAll()
        {
            return _contactsRepository.GetAll().Select(c => c.ToDTO()).ToList();
        }

        public ContactDTO? GetById(Guid id)
        {
            try
            {
                var found = _contactsRepository.GetById(id);
                return found.ToDTO();
            }
            catch (KeyNotFoundException)
            {
                return null;
            }
        }

        public Guid Update(Guid id, UpdateContactDTO updateContactDto)
        {
            try
            {
                var contact = ContactDTO.ToEntity(
                    new ContactDTO
                    {
                        Id = id,
                        Name = updateContactDto.Name,
                        Surname = updateContactDto.Surname,
                        Email = updateContactDto.Email,
                        Password = updateContactDto.Password,
                        CategoryId = updateContactDto.CategoryId,
                        SubcategoryId = updateContactDto.SubcategoryId,
                        CustomSubCategory = updateContactDto.CustomSubCategory,
                        Phone = updateContactDto.Phone,
                        BirthDate = updateContactDto.BirthDate
                    }
                    );
                if (contact == null)
                {
                    return Guid.Empty;
                }
                _contactsRepository.Update(contact);
                return contact.Id;
            }
            catch (ArgumentException) { return Guid.Empty; }
            catch (KeyNotFoundException) { return Guid.Empty; }
        }
    }
}

File: Frontend/src/app/services/contacts.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { JwtHelperService } from '@auth0/angular-jwt';
import { Observable, throwError, forkJoin, of } from 'rxjs';
import { catchError, map, switchMap } from 'rxjs/operators';
import { environment } from '../../environments/environment';
import { ContactModel, ContactDto, CreateContactDto, UpdateContactDto } from '../models/contact.model';
import { CategoryDto } from '../models/category.model'
import { CategoriesService } from '../services/categories.service';
import { SubCategoriesService } from './subcategories.service';

@Injectable({
  providedIn: 'root',
})
export class ContactsService {
  private base = `${environment.apiUrl}/contacts`;
  private jwtHelper = new JwtHelperService();

  constructor(private http: HttpClient,
    private categoriesService: CategoriesService,
    private subcategoriesService: SubCategoriesService) { }

  private authOptions() {
    const token = localStorage.getItem('token');
    if (!token) return {};
    return { headers: { Authorization: `Bearer ${token}` } };
  }

  public getAll(): Observable<ContactModel[]> {
    return this.http.get<ContactDto[]>(this.base).pipe(
      switchMap((dtos: ContactDto[]) => {
        if (dtos.length === 0) return of([]);
        const observables = dtos.map(dto => this.mapDtoToContact(dto));
        return forkJoin(observables);
      }),
      catchError((error: HttpErrorResponse) => throwError(() => new Error('Something went wrong')))
    );
  }

  public getById(id: string): Observable<ContactModel> {
    return this.http.get<ContactDto>(`${this.base}/${id}`).pipe(
      switchMap((dto: ContactDto) => this.mapDtoToContact(dto)),
      catchError((error: HttpErrorResponse) => throwError(() => new Error('Something went wrong')))
    );
  }

  public create(contact: CreateContactDto): Observable<CreateContactDto> {
    return this.http.post<CreateContactDto>(this.base, contact, this.authOptions()).pipe(
      catchError((error: HttpErrorResponse) => throwError(() => new Error('Something went wrong'))
    ));
  }

  public update(id: string, contact: UpdateContactDto): Observable<UpdateContactDto> {
    return this.http.put<any>(`${this.base}/${id}`, contact, this.authOptions()).pipe(
      catchError((error: HttpErrorResponse) => throwError(() => new Error('Something went wrong'))
    ));
  }

  public delete(id: string): Observable<ContactDto> {
    return this.http.delete<ContactDto>(`${this.base}/${id}`, this.authOptions()).pipe(
      catchError((error: HttpErrorResponse) => throwError(() => new Error('Something went wrong'))
    ));
  }

  private mapDtoToContact(dto: ContactDto): Observable<ContactModel> {
    if (dto.categoryId != null && dto.subcategoryId != null) {
      return forkJoin({
        category: this.categoriesService.getById(dto.categoryId),
        subcategory: this.subcategoriesService.getById(dto.subcategoryId)
      }).pipe(
        map(({ category, subcategory }) => {
          const contact: ContactModel = {
            id: dto.id,
            name: dto.name,
            surname: dto.surname,
            email: dto.email,
            password: dto.password,
            phone: dto.phone,
            birthDate: dto.birthDate,
            category: category.name,
            subcategory: subcategory.name
          };
          return contact;
        })
      );
    } else {
      const contact: ContactModel = {
            id: dto.id,
            name: dto.name,
            surname: dto.surname,
            email: dto.email,
            password: dto.password,
            phone: dto.phone,
            birthDate: dto.birthDate,
            category: 'inne',
            subcategory: dto.customSubCategory
          };
      return of(contact);
    }
  }
}
